modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+                ],
+                self.w2[0],
+            )
+            + torch.einsum(
+                "nyxbi,bio->nyxbo",
+                o1_real[
+                    :, total_modes - kept_modes : total_modes + kept_modes, :kept_modes
+                ],
+                self.w2[1],
+            )
+            + self.b2[1]
+        )
+
+        x = F.softshrink(o2, lambd=self.sparsity_threshold)
+        x = fft.view_as_complex(x)
+        # TODO(akamenev): replace the following branching with
+        # a one-liner, something like x.reshape(..., -1).squeeze(-1),
+        # but this currently fails during ONNX export.
+        if torch.onnx.is_in_onnx_export():
+            x = x.reshape(B, H, W // 2 + 1, C, 2)
+        else:
+            x = x.reshape(B, H, W // 2 + 1, C)
+        # Using ONNX friendly FFT functions
+        x = fft.irfft2(x, s=(H, W), dim=(1, 2), norm="ortho")
+        x = x.type(dtype)
+
+        return x + bias
+
+
+class Block(nn.Module):
+    """AFNO block, spectral convolution and MLP
+
+    Parameters
+    ----------
+    embed_dim : int
+        Embedded feature dimensionality
+    mod_dim : int
+        Modululation input dimensionality
+    num_blocks : int, optional
+        Number of blocks used in the block diagonal weight matrix, by default 8
+    mlp_ratio : float, optional
+        Ratio of MLP latent variable size to input feature size, by default 4.0
+    drop : float, optional
+        Drop out rate in MLP, by default 0.0
+    activation_fn: nn.Module, optional
+        Activation function used in MLP, by default nn.GELU
+    norm_layer : nn.Module, optional
+        Normalization function, by default nn.LayerNorm
+    double_skip : bool, optional
+        Residual, by default True
+    sparsity_threshold : float, optional
+        Sparsity threshold (softshrink) of spectral features, by default 0.01
+    hard_thresholding_fraction : float, optional
+        Threshold for limiting number of modes used [0,1], by default 1
+    modulate_filter: bool, optional
+        Whether to compute the modulation for the FFT filter
+    modulate_mlp: bool, optional
+        Whether to compute the modulation for the MLP
+    scale_shift_mode: str
+        If 'complex' (default), compute the scale-shift operation using complex
+        operations. If 'real', use real operations.
+    """
+
+    def __init__(
+        self,
+        embed_dim: int,
+        mod_dim: int,
+        num_blocks: int = 8,
+        mlp_ratio: float = 4.0,
+        drop: float = 0.0,
+        activation_fn: nn.Module = nn.GELU(),
+        norm_layer: nn.Module = nn.LayerNorm,
+        double_skip: bool = True,
+        sparsity_threshold: float = 0.01,
+        hard_thresholding_fraction: float = 1.0,
+        modulate_filter: bool = True,
+        modulate_mlp: bool = True,
+        scale_shift_mode: str = "real",
+    ):
+        super().__init__()
+        self.norm1 = norm_layer(embed_dim)
+        if modulate_filter:
```

Comment:
Would users realistically want to shut this off? Would this just become AFNO with out these? I wonder if this option is really needed or can just be considered part of the arch.

Reply:
When developing, I was testing this with either `modulate_mlp` and `modulate_filter` turned off. It turns out to work best with both, but it was surprisingly good with only `modulate_mlp` active. Anyway, this should be handled without conditionals in the forward now.